log4j-core/src/main/java/org/apache/logging/log4j/core/appender/AsyncAppenderEventForwarder.java

```diff
@@ -0,0 +1,147 @@
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the license for the specific language governing permissions and
+ * limitations under the license.
+ */
+package org.apache.logging.log4j.core.appender;
+
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.core.LogEvent;
+import org.apache.logging.log4j.core.config.AppenderControl;
+import org.apache.logging.log4j.core.util.Log4jThread;
+import org.apache.logging.log4j.status.StatusLogger;
+
+import java.util.List;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.atomic.AtomicLong;
+
+class AsyncAppenderEventForwarder extends Log4jThread {
+
+    private static final AtomicLong THREAD_COUNTER = new AtomicLong(0);
+
+    private static final Logger LOGGER = StatusLogger.getLogger();
+
+    private final AppenderControl errorAppender;
+
+    private final List<AppenderControl> appenders;
+
+    private final BlockingQueue<LogEvent> queue;
+
+    private volatile boolean stopped;
+
+    AsyncAppenderEventForwarder(
+            final String name,
+            final AppenderControl errorAppender,
+            final List<AppenderControl> appenders,
+            final BlockingQueue<LogEvent> queue) {
+        super("AsyncAppenderEventForwarder-" + THREAD_COUNTER.incrementAndGet() + "-" + name);
+        this.errorAppender = errorAppender;
+        this.appenders = appenders;
+        this.queue = queue;
+        this.stopped = false;
+    }
+
+    @Override
+    public void run() {
+        LOGGER.trace("{} has started.", getName());
+        forwardAll();
+        forwardRemaining();
+    }
+
+    private void forwardAll() {
+        while (!stopped) {
+            LogEvent event;
+            try {
+                event = queue.take();
+            } catch (final InterruptedException ignored) {
+                // Restore the interrupted flag cleared when the exception is caught.
+                interrupt();
+                break;
+            }
+            event.setEndOfBatch(queue.isEmpty());
+            forwardOne(event);
+        }
+        LOGGER.trace("{} has stopped.", getName());
+    }
+
+    private void forwardRemaining() {
+        int eventCount = 0;
+        while (true) {
+            // Note the non-blocking Queue#poll() method!
+            final LogEvent event = queue.poll();
+            if (event == null) {
+                break;
+            }
+            event.setEndOfBatch(queue.isEmpty());
+            forwardOne(event);
+            eventCount++;
+        }
+        LOGGER.trace(
+                "{} has processed the last {} remaining event(s).",
+                getName(), eventCount);
+    }
+
+    void forwardOne(final LogEvent event) {
```

Comment:
I would rename to "forward" since you give the method one argument always and it's not like we have to distinguish with a forwardTwo method ;-) or am I missing something?

Reply:
+1 I don't think forward is correct here or in the class name.

Reply:
Oops, I misread your comment. I think it's odd that this object is both a thread, and called a forwarder. I think that made sense before, but the background thread is a receiver, and this provides functionality to submit events. Naming things is one of the two hardest problems in CS, and I don't have better alternatives in mind.

To be clear I'm not suggesting it must be renamed, just thinking aloud.

Reply:
@garydgregory, okay, I will rename it to `forward()`.

@carterkozak, what about *dispatcher*?

Reply:
For the records, with `forwardOne()`, I was trying to stick to the Spring Data conventions: `findOne()`, `findAll()`, etc.

Reply:
I am not a fan of following Spring's naming, we need to pick names that make sense FOR LOG4J ;-)

Reply:
Renamed to `forward()`.

Reply:
`dispatcher` or `handler`. No strong opinion, I'm terrible at naming.

Reply:
Renamed the class to `AsyncAppenderEventDispatcher`.